flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,213 @@
+                    currentReader);
+            if (currentSourceIndex + 1 < realReaders.size()) {
+                // Signal the coordinator that the current reader has consumed all input and the
+                // next source can potentially be activated (after all readers are ready).
+                readerContext.sendSourceEventToCoordinator(
+                        new SourceReaderFinishedEvent(currentSourceIndex, lastCheckpointId));
+                // More data will be available from the next reader.
+                // InputStatus.NOTHING_AVAILABLE requires us to complete the availability
+                // future after source switch to resume poll.
+                return InputStatus.NOTHING_AVAILABLE;
+            }
+        }
+        return status;
+    }
+
+    @Override
+    public List<HybridSourceSplit<?>> snapshotState(long checkpointId) {
+        this.lastCheckpointId = checkpointId;
+        List<? extends SourceSplit> state = currentReader.snapshotState(checkpointId);
+        return wrappedSplits(currentSourceIndex, state);
+    }
+
+    public static List<HybridSourceSplit<?>> wrappedSplits(
+            int readerIndex, List<? extends SourceSplit> state) {
+        List<HybridSourceSplit<?>> wrappedSplits = new ArrayList<>(state.size());
+        for (SourceSplit split : state) {
+            wrappedSplits.add(new HybridSourceSplit<>(readerIndex, split));
+        }
+        return wrappedSplits;
+    }
+
+    public static <SplitT extends SourceSplit> List<SplitT> unwrappedSplits(
+            List<HybridSourceSplit<SplitT>> splits) {
+        List<SplitT> unwrappedSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit<SplitT> split : splits) {
+            unwrappedSplits.add(split.getWrappedSplit());
+        }
+        return unwrappedSplits;
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return availabilityFuture = currentReader.isAvailable();
+    }
+
+    @Override
+    public void addSplits(List<HybridSourceSplit<?>> splits) {
+        LOG.info(
+                "Adding splits subtask={} sourceIndex={} {} {}",
+                readerContext.getIndexOfSubtask(),
+                currentSourceIndex,
+                currentReader,
+                splits);
+        List<SourceSplit> realSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit<?> split : splits) {
+            Preconditions.checkState(
+                    split.sourceIndex() == currentSourceIndex,
+                    "Split %s while current source is %s",
+                    split,
+                    currentSourceIndex);
+            realSplits.add(split.getWrappedSplit());
+        }
+        currentReader.addSplits((List) realSplits);
+    }
+
+    @Override
+    public void notifyNoMoreSplits() {
+        currentReader.notifyNoMoreSplits();
+        LOG.debug(
+                "No more splits for reader subtask={} sourceIndex={} {}",
```

Comment:
Could the log message be `No more splits for reader subtask={} sourceIndex={} currentReader={}`?